lua/nvim-treesitter/indent.lua

```diff
@@ -52,52 +61,59 @@ local function find_delimiter(bufnr, node, delimiter)
   end
 end
 
-local get_indents = tsutils.memoize_by_buf_tick(function(bufnr, root, lang)
-  local map = {
-    indent = {
-      auto = {},
-      begin = {},
-      ["end"] = {},
-      dedent = {},
-      branch = {},
-      ignore = {},
-      align = {},
-      zero = {},
-    },
-  }
-
-  local function split(to_split)
-    local t = {}
-    for str in string.gmatch(to_split, "([^.]+)") do
-      table.insert(t, str)
+---Memoize a function using hash_fn to hash the arguments.
+---@generic F: function
+---@param fn F
+---@param hash_fn fun(...): any
+---@return F
+local function memoize(fn, hash_fn)
+  local cache = setmetatable({}, { __mode = "kv" }) ---@type table<any,any>
+
+  return function(...)
+    local key = hash_fn(...)
+    if cache[key] == nil then
+      local v = fn(...) ---@type any
+      cache[key] = v ~= nil and v or vim.NIL
     end
-    return t
+
+    local v = cache[key]
+    return v ~= vim.NIL and v or nil
   end
+end
+
+local get_indents = memoize(function(bufnr, root, lang)
+  local map = {
+    ["indent.auto"] = {},
+    ["indent.begin"] = {},
+    ["indent.end"] = {},
+    ["indent.dedent"] = {},
+    ["indent.branch"] = {},
+    ["indent.ignore"] = {},
+    ["indent.align"] = {},
+    ["indent.zero"] = {},
+  }
 
-  for name, node, metadata in queries.iter_captures(bufnr, "indents", root, lang) do
-    local path = split(name)
-    -- node may contain a period so append directly.
-    table.insert(path, node:id())
-    queries.insert_to_path(map, path, metadata or {})
+  --TODO(clason): remove when dropping Nvim 0.8 compat
+  local query = (ts.query.get or ts.get_query)(lang, "indents")
+  for id, node, metadata in query:iter_captures(root, bufnr) do
+    map[query.captures[id]][node:id()] = metadata or {}
   end
 
   return map
-end, {
-  -- Memoize by bufnr and lang together.
-  key = function(bufnr, root, lang)
-    return tostring(bufnr) .. root:id() .. "_" .. lang
-  end,
-})
+end, function(bufnr, root, lang)
+  return tostring(bufnr) .. root:id() .. "_" .. lang
+end)
 
 ---@param lnum number (1-indexed)
 function M.get_indent(lnum)
   local bufnr = vim.api.nvim_get_current_buf()
-  local parser = parsers.get_parser(bufnr)
+  local parser = ts.get_parser(bufnr)
   if not parser or not lnum then
     return -1
   end
 
-  local root_lang = parsers.get_buf_lang(bufnr)
+  local ft = vim.bo[bufnr].filetype
+  local root_lang = vim.treesitter.language.get_lang(ft) or ft
```

Comment:
Im pretty sure `get_lang(ft)` changes have yet to be backported, so this will be a regression to some fts (if any)

Reply:
Aside from this, I agree with the changes.
Should this also be changed for `memoize_by_buf_tick` directly, as `nvim-treesitter#foldexpr()` still relies on it?

Reply:
I wouldn't touch `foldexpr`, as it's already obsolete -- `vim.treesitter.foldexpr` was added in 0.9.0.